src/compiler/moduleNameResolver.ts

```diff
@@ -18,7 +18,8 @@ namespace ts {
                 packageId = {
                     name: packageJsonContent.name,
                     subModuleName: r.path.slice(packageInfo.packageDirectory.length + directorySeparator.length),
-                    version: packageJsonContent.version
+                    version: packageJsonContent.version,
+                    bundlesTypes: typeof packageJsonContent.types === "string" || typeof packageJsonContent.typings === "string"

```

Comment:
The problem is that root `index.d.ts` files implicitly count, even if `types` or `typings` isn’t set. I think instead of changing `PackageId`, take a look at how `typesPackageExists` works in the function you changed in checker.ts, then follow it to the implementation of `getPackagesSet`, which is _only_ used in `typesPackageExists`.

```ts
const getPackagesSet = memoize(() => {
    const set = new Set<string>();
    host.getSourceFiles().forEach(sf => {
        if (!sf.resolvedModules) return;

        forEachEntry(sf.resolvedModules, r => {
            if (r && r.packageId) set.add(r.packageId.name);
        });
    });
    return set;
});
```

The resolved module `r` has an `extension` field. You could change this Set to a Map whose values are true/false based on whether you saw any resolved modules from that package with .d.ts extensions. It’s not perfect because only modules that are referenced somewhere in the program will show up, but I think that’s close enough. It’s also nice because it will catch non-index `.d.ts` files for the package, like for a package that has no root typings file for some reason but has all its sub-modules typed.

Don’t take this as the only solution; there are almost certainly other good ways to go about this.